File: app/src/main/java/com/saienko/androidthings/barman/ui/CocktailActivity.java
package com.saienko.androidthings.barman.ui;

import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.Color;
import android.os.Bundle;
import android.util.Log;
import android.util.LongSparseArray;
import android.widget.Button;
import com.saienko.androidthings.barman.R;
import com.saienko.androidthings.barman.db.cocktail.Cocktail;
import com.saienko.androidthings.barman.db.cocktail.CocktailElement;
import com.saienko.androidthings.barman.listener.OnResultListener;
import com.saienko.androidthings.barman.listener.OnUpdateListener;
import com.saienko.androidthings.barman.receiver.CocktailBroadcastReceiver;
import com.saienko.androidthings.barman.receiver.CocktailUpdateBroadcastReceiver;
import com.saienko.androidthings.barman.service.CocktailService;
import com.saienko.androidthings.barman.ui.base.BaseActivity;
import devlight.io.library.ArcProgressStackView;

import java.util.ArrayList;

public class CocktailActivity extends BaseActivity {

    private static final String TAG = "CocktailActivity";

    private static final String EXTRA_COCKTAIL = "EXTRA_COCKTAIL";
    private static final long   TOTAL_POSITION = -1;

    CocktailBroadcastReceiver       cocktailBroadcastReceiver;
    CocktailUpdateBroadcastReceiver cocktailUpdateBroadcastReceiver;

    Cocktail cocktail;

    LongSparseArray<ArcProgressStackView.Model> holderMap;
    LongSparseArray<Boolean>                    cocktailEnd;
    LongSparseArray<Integer>                    progressArray;
    private Button               btnStart;
    private Button               btnStop;
    private ArcProgressStackView mArcProgressStackView;
    private int                  modelCount;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_cocktail);
        holderMap = new LongSparseArray<>();
        progressArray = new LongSparseArray<>();
        cocktailEnd = new LongSparseArray<>();
        handleIntent();
        registerBroadcastReceiver();
        initUI();
    }

    private void handleIntent() {
        cocktail = getIntent().getParcelableExtra(EXTRA_COCKTAIL);
    }

    public static void start(Context context, Cocktail cocktail) {
        Intent intent = new Intent(context, CocktailActivity.class);
        intent.putExtra(EXTRA_COCKTAIL, cocktail);
        context.startActivity(intent);
    }

    @Override
    public void initUI() {
        super.initUI();
        modelCount = cocktail.getCocktailElements().size() + 1;
        btnStart = findViewById(R.id.btnStart);
        btnStop = findViewById(R.id.btnStop);
        mArcProgressStackView = findViewById(R.id.apsv);

        final int[] startColors = getResources().getIntArray(R.array.green);
//        final int[] endColors   = getResources().getIntArray(R.array.orange);
        final int[] bgColors = getResources().getIntArray(R.array.grey);

        final ArrayList<ArcProgressStackView.Model> models = new ArrayList<>();
        for (int i = 0; i < cocktail.getCocktailElements().size(); i++) {
            CocktailElement element =
                    cocktail.getCocktailElements().get(i);
            if (element.getPosition() == null || element.getPosition().getMotor() == null) {
                break;
            }

            cocktailEnd.put(element.getPosition().getMotor().getGpioId(), false);
            ArcProgressStackView.Model model =
                    new ArcProgressStackView.Model(element.getComponent().getName(), i, bgColors[i],
                                                   startColors[i]);
            models.add(model);
            holderMap.put(element.getPosition().getMotor().getGpioId(), model);
            progressArray.put(element.getPosition().getMotor().getGpioId(), 0);
        }
        ArcProgressStackView.Model totalModel =
                new ArcProgressStackView.Model("Total", cocktail.getCocktailElements().size(),
                                               bgColors[cocktail.getCocktailElements().size()],
                                               startColors[cocktail.getCocktailElements().size()]
                );
        holderMap.put(TOTAL_POSITION, totalModel);
        progressArray.put(TOTAL_POSITION, 0);
        models.add(totalModel);

        mArcProgressStackView.setModels(models);

        btnStart.setOnClickListener(view -> {
            CocktailService.startCocktail(CocktailActivity.this, cocktail);
            setProgressValue(0);
        });
        btnStop.setOnClickListener(view -> CocktailService.stopCocktail(CocktailActivity.this));

        mArcProgressStackView.setShadowColor(Color.argb(200, 0, 0, 0));
//        mArcProgressStackView.setAnimationDuration(1000);
//        mArcProgressStackView.setSweepAngle(270);

    }

    private void registerBroadcastReceiver() {
        cocktailUpdateBroadcastReceiver = new CocktailUpdateBroadcastReceiver(new OnUpdateListener() {
            @Override
            public void onItemUpdate(long gpioId, int progress) {
                Log.d(TAG, "onItemUpdate: gpioId" + gpioId + " progress=" + progress);
                setValue(gpioId, progress);
            }

            @Override
            public void onItemFinish(long gpioId) {
                Log.d(TAG, "onItemFinish: " + gpioId);
                calculateFinish(gpioId);
            }

            @Override
            public void onItemStart(long gpioId) {
                Log.d(TAG, "onItemStart: " + gpioId);
                btnStart.setEnabled(false);
                btnStop.setEnabled(true);
//                setProgressValue(0);
            }
        });
        cocktailBroadcastReceiver = new CocktailBroadcastReceiver(new OnResultListener() {

            @Override
            public void onError() {
                Log.d(TAG, "onError: ");
                btnStart.setEnabled(false);
                btnStop.setEnabled(false);
            }

            @Override
            public void onCancel() {
                Log.d(TAG, "onCancel: ");
                btnStart.setEnabled(true);
                btnStop.setEnabled(false);
            }
        });
        IntentFilter intentFilter = new IntentFilter(CocktailService.BROADCAST_COCKTAIL_STATUS);
        intentFilter.addCategory(Intent.CATEGORY_DEFAULT);
        registerReceiver(cocktailBroadcastReceiver, intentFilter);

        IntentFilter intentUpdateFilter = new IntentFilter(CocktailService.BROADCAST_COCKTAIL_PROGRESS);
        intentUpdateFilter.addCategory(Intent.CATEGORY_DEFAULT);
        registerReceiver(cocktailUpdateBroadcastReceiver, intentUpdateFilter);
    }

    private void setProgressValue(int progress) {
        for (ArcProgressStackView.Model model : mArcProgressStackView.getModels()) {
            model.setProgress(progress);
        }
        mArcProgressStackView.postInvalidate();

        for (int i = 1; i < modelCount; i++) {
            progressArray.put(progressArray.keyAt(i), progress);
        }
    }

    private void setValue(long gpioId, int progress) {
        ArcProgressStackView.Model model = holderMap.get(gpioId);
        if ((int) model.getProgress() != progress) {
            model.setProgress(progress);
        }
        progressArray.put(gpioId, progress);

        ArcProgressStackView.Model totalModel    = holderMap.get(TOTAL_POSITION);
        float                      totalProgress = getTotalProgress();
        if (totalModel.getProgress() != totalProgress) {
            totalModel.setProgress(totalProgress);
        }
        mArcProgressStackView.postInvalidate();
    }

    private float getTotalProgress() {
        int total = 0;
        for (int i = 1; i < modelCount; i++) {
            if (progressArray.keyAt(i) != TOTAL_POSITION) {
                if (progressArray.valueAt(i) != null) {
                    total += progressArray.valueAt(i);
                }

            }
        }
        return total / (modelCount - 1);
    }

    private void calculateFinish(long gpioId) {
        cocktailEnd.put(gpioId, true);
        for (int i = 0; i < cocktailEnd.size(); i++) {
            long    key   = cocktailEnd.keyAt(i);
            boolean value = cocktailEnd.get(key);
            if (!value) {
                return;
            }
        }
        btnStart.setEnabled(true);
        btnStop.setEnabled(false);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        unregisterReceiver(cocktailBroadcastReceiver);
        unregisterReceiver(cocktailUpdateBroadcastReceiver);
    }
}

File: app/src/main/java/com/saienko/androidthings/barman/ui/CocktailActivity.kt
package com.saienko.androidthings.barman.ui

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.util.LongSparseArray
import android.view.View
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TextView
import com.saienko.androidthings.barman.R
import com.saienko.androidthings.barman.db.entity.Cocktail
import com.saienko.androidthings.barman.db.manager.CocktailManager
import com.saienko.androidthings.barman.listener.OnResultListener
import com.saienko.androidthings.barman.listener.OnUpdateListener
import com.saienko.androidthings.barman.receiver.CocktailBroadcastReceiver
import com.saienko.androidthings.barman.receiver.CocktailUpdateBroadcastReceiver
import com.saienko.androidthings.barman.service.CocktailService
import com.saienko.androidthings.barman.ui.adapter.CocktailElementAdapter
import com.saienko.androidthings.barman.ui.base.BaseActivity
import com.saienko.androidthings.barman.ui.view.ProgressView
import kotlinx.coroutines.experimental.CommonPool
import kotlinx.coroutines.experimental.android.UI
import kotlinx.coroutines.experimental.async
import kotlinx.coroutines.experimental.launch


class CocktailActivity : BaseActivity() {

    private lateinit var cocktailBroadcastReceiver: CocktailBroadcastReceiver
    private lateinit var cocktailUpdateBroadcastReceiver: CocktailUpdateBroadcastReceiver

    private lateinit var cocktail: Cocktail

    private var holderMap: LongSparseArray<ProgressView>? = null
    private var cocktailEnd: LongSparseArray<Boolean>? = null
    private var progressArray: LongSparseArray<Int>? = null

    private lateinit var btnStart: Button
    private lateinit var btnStop: Button
    private lateinit var cocktailItemList: RecyclerView
    private lateinit var progressContainer: LinearLayout
    private lateinit var errorView: TextView

    private var cocktailId: Long = -1

    private var modelCount: Int = 0

    private val totalProgress: Int
        get() {
            val total = (1..modelCount)
                    .filter { progressArray!!.keyAt(it) != TOTAL_POSITION && progressArray!!.valueAt(it) != null }
                    .sumBy { progressArray!!.valueAt(it) }
            return total / (modelCount - 1)
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_cocktail)
        holderMap = LongSparseArray()
        progressArray = LongSparseArray()
        cocktailEnd = LongSparseArray()
        handleIntent()
        registerBroadcastReceiver()
        initUI()
    }

    override fun onResume() {
        super.onResume()
        getCocktail(cocktailId)
    }

    private fun handleIntent() {
        cocktailId = intent.getLongExtra(EXTRA_COCKTAIL_ID, -1)
    }

    private fun getCocktail(cocktailId: Long) {
        val cocktailManager = CocktailManager()
        launch(UI) {
            initCocktail(async(CommonPool) { cocktailManager.get(cocktailId) }.await())
        }
    }

    private fun initCocktail(cocktail: Cocktail) {
        this.cocktail = cocktail
        if (!cocktailPositionsCorrect(cocktail)) {
            blockUI("Positions are not correct")
        } else if (!cocktailComponentsCorrect(cocktail)) {
            blockUI("Components are not correct")
        } else {
            updateUi(cocktail)
        }
    }

    private fun blockUI(text: String) {
        btnStart.isEnabled = false
        btnStop.isEnabled = false
        errorView.visibility = View.VISIBLE
        errorView.text = text
        progressContainer.visibility = View.GONE
    }

    private fun cocktailComponentsCorrect(cocktail: Cocktail): Boolean {
        return cocktail.cocktailElements!!.none { it.component == null }
    }

    private fun cocktailPositionsCorrect(cocktail: Cocktail): Boolean {
        return cocktail.cocktailElements!!.none { it.position == null }
    }

    private fun updateUi(cocktail: Cocktail) {
        progressContainer.visibility = View.VISIBLE
        errorView.visibility = View.GONE
        modelCount = cocktail.cocktailElements!!.size + 1
        cocktail.cocktailElements!!.forEachIndexed { _, element ->
            if (element.position == null || element.position!!.motor == null) {
                return@forEachIndexed
            }

            cocktailEnd!!.put(element.position!!.motor!!.gpioId, false)
            progressArray!!.put(element.position!!.motor!!.gpioId, 0)

            val progressView = ProgressView(this)
            progressView.setName(element.component!!.componentName)
            holderMap!!.put(element.position!!.motor!!.gpioId, progressView)
            progressContainer.addView(progressView)
        }
        val totalView = ProgressView(this)
        totalView.setName("Total")
        holderMap!!.put(TOTAL_POSITION, totalView)
        progressArray!!.put(TOTAL_POSITION, 0)
        progressContainer.addView(totalView)
        showCocktailItemList(cocktail)
    }

    private fun showCocktailItemList(cocktail: Cocktail) {
        val list = CocktailElementAdapter()
        val layoutManager = LinearLayoutManager(this)
        cocktailItemList.layoutManager = layoutManager
        cocktail.cocktailElements?.let {
            list.addItems(it)
            cocktailItemList.adapter = list
        }
    }

    override fun initUI() {
        super.initUI()
        btnStart = findViewById(R.id.btnStart)
        btnStop = findViewById(R.id.btnStop)
        cocktailItemList = findViewById(R.id.cocktailItemList)
        progressContainer = findViewById(R.id.progressContainer)
        errorView = findViewById(R.id.errorView)

        btnStart.setOnClickListener {
            CocktailService.startCocktail(this@CocktailActivity, cocktail.id)
            setProgressValue(0)
        }
        btnStop.setOnClickListener { CocktailService.stopCocktail(this@CocktailActivity) }
    }

    private fun registerBroadcastReceiver() {
        cocktailUpdateBroadcastReceiver = CocktailUpdateBroadcastReceiver(object : OnUpdateListener {
            override fun onItemUpdate(gpioId: Long, progress: Int) {
                setValue(gpioId, progress)
            }

            override fun onItemFinish() {
                btnStart.isEnabled = true
                btnStop.isEnabled = false
            }

            override fun onItemStart() {
                btnStart.isEnabled = false
                btnStop.isEnabled = true
            }
        })
        cocktailBroadcastReceiver = CocktailBroadcastReceiver(object : OnResultListener {

            override fun onError() {
                Log.d(TAG, "onError: ")
                btnStart.isEnabled = false
                btnStop.isEnabled = false
            }

            override fun onCancel() {
                Log.d(TAG, "onCancel: ")
                btnStart.isEnabled = true
                btnStop.isEnabled = false
            }
        })
        val intentFilter = IntentFilter(CocktailService.BROADCAST_COCKTAIL_STATUS)
        intentFilter.addCategory(Intent.CATEGORY_DEFAULT)
        registerReceiver(cocktailBroadcastReceiver, intentFilter)

        val intentUpdateFilter = IntentFilter(CocktailService.BROADCAST_COCKTAIL_PROGRESS)
        intentUpdateFilter.addCategory(Intent.CATEGORY_DEFAULT)
        registerReceiver(cocktailUpdateBroadcastReceiver, intentUpdateFilter)
    }

    private fun setProgressValue(progress: Int) {
        this.progressContainer.postInvalidate()
        for (i in 1..modelCount) {
            progressArray!!.put(progressArray!!.keyAt(i), progress)
        }
    }

    private fun setValue(gpioId: Long, progress: Int) {
        val progressView = holderMap!!.get(gpioId)
        if (progressView != null) {
            if (progressView.progress != progress) {
                progressView.progress = progress
            }
            progressArray!!.put(gpioId, progress)

            val totalView = holderMap!!.get(TOTAL_POSITION)
            val totalProgress = totalProgress
            if (totalView.progress != totalProgress) {
                totalView.progress = totalProgress
            }
            this.progressContainer.postInvalidate()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        unregisterReceiver(cocktailBroadcastReceiver)
        unregisterReceiver(cocktailUpdateBroadcastReceiver)
    }

    companion object {

        private const val TAG = "CocktailActivity"
        private const val EXTRA_COCKTAIL_ID = "EXTRA_COCKTAIL_ID"
        private const val TOTAL_POSITION: Long = -1

        fun start(context: Context, cocktailId: Long) {
            val intent = Intent(context, CocktailActivity::class.java)
            intent.putExtra(EXTRA_COCKTAIL_ID, cocktailId)
            context.startActivity(intent)
        }
    }
}
